Treat all warnings as errors

hctrdev opened the pull request:
## Description

Update project definition to treat all warnings as errors. If the compiler finds a warning it should not be ignored. Treating it as an error ensures this is the case.

## TODO

- [x] Update CHANGELOG

Commit:
Treat all warnins as errors

Update the project definition to treat all warnings as errors in the main
project. Warnings are still allowed in the test project.
Fix errors in the main project.
Add HashCode method to the multi value map class.
Split multi value map interface into separate file.
Reset library version to `0.0.0.0` as it is updated automatically by the
build script.

Commit:
Code cleanup to fix some compiler info messages

Update private properties to be read-only where possible.
Inline variable definitions where possible.
Use `[^x]` array indexing instead of `[length - x]` based on C#
recommendation.

Commit:
Update changelog

ColmBhandal commented on CsharpExtras/CsharpExtras.csproj:
```diff
@@ -1,11 +1,12 @@
-	<TargetFramework>netstandard2.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <LangVersion>8.0</LangVersion>
+    <WarningsAsErrors></WarningsAsErrors>
```
Can you get rid of this empty XML element now?

ColmBhandal commented on CsharpExtras/Dictionary/MultiValueMapImpl.cs:
```diff
@@ -168,5 +154,13 @@ public override bool Equals(object obj)
             return true;
         }
+
+        public override int GetHashCode()
```
This algorithm isn't immediately clear. Is it a standard practice? Can we add some commenting around it to explain what's going on?

ColmBhandal commented on CsharpExtras/Extensions/CollectionExtension.cs:
```diff
@@ -8,7 +8,7 @@ namespace CsharpExtras.Extensions
     public static class CollectionExtension
     {
-        public static bool TryGetSingleton<TValueType>(this ICollection<TValueType> collection, out TValueType singleton)
+        public static bool TryGetSingleton<TValueType>(this ICollection<TValueType> collection, out TValueType? singleton) where TValueType : class
```
Technically, this is a breaking change. Consider adding this change to the change log file. The change log entry needn't mention this class directly- it could just say that some generic constraints have been added. Or you could say "some constraints have been added e.g." and then give this as an example.

ColmBhandal replied:
I advise to rename ``TValueType`` to something like ``TVal``. "Value type" in C# has a very specific meaning and we don't want to confuse people; particularly since the new  type constraint enforces that the type is _not_ a value type. I'm aware that this wasn't a change you made in this PR but I'm just noticing it now.

ColmBhandal reviewed (changes requested):
See comments

Commit:
Add test for the new GetHashCode method

Add tests for the new GetHashCode method to ensure that two maps that are
equal will return the same hashcode.
Fix incorrect implementation of the GetHashCode function. The new
implementation iterates over the keys, then over all the values and XORs
all the hash codes together.
Add test to ensure no exceptions are thrown on large maps when calculating
the hash code (e.g. Overflow exceptions).

ColmBhandal commented on CsharpExtrasTest/Dictionary/MultiValueMapTest.cs:
```diff
@@ -95,5 +95,51 @@ public void GivenMap_WhenTransform_ThenResultIsOfCorrectLengthAndContainsExpecte
+            map1.Add(2, "d");
+            map1.Add(2, "e");
+
+            map2.Add(2, "d");
```
The order of d and e is still the same here as above, although interleaved with other statements. I'd swap them.

ColmBhandal commented on CsharpExtras/Dictionary/MultiValueMapImpl.cs:
```diff
@@ -168,5 +154,25 @@ public override bool Equals(object obj)
             return true;
         }
+
+        public override int GetHashCode()
```
The issue with this is it becomes quite inefficient. We have a hashcode generation function that's linear in the number of elements in all the value sets. Considering the hash code is used to store elements in things like hash sets, which are supposed to be efficient, this is problematic.

Perhaps a better solution is to remove the ``equals`` override and instead add another equality method e.g. ``mapEquals`` that's analogoug to the C# function ``setEquals`` that's in the set interface.

hctrdev replied:
I think getting a hashcode is always going to be inefficient. If you are calculating a hash for any data type (string, list, set, etc) you have to iterate over the whole thing at least once to calculate the hash. With equality you can sometimes break early, but not when calculating a hash.

There is another issue with the hashcode though that we should consider: [stackoverflow question](https://stackoverflow.com/questions/10567450/implement-gethashcode-for-objects-that-contain-collections). The idea is that if you setup your hashcode definition to change when the data in the object changes, and if your object happens to be in a set or dictionary, it can get lost.

ColmBhandal reviewed (changes requested):
Still some issues to go over, particularly with hash code

Commit:
Implement PR feedback

Rename generic type on the collection extension methods to avoid confusion
with the C# definition of "value type".
Re-order elements in the multi-value map during testing.
Add changelog entry for breaking change to the `TryGetSingleton`.
